llpc/context/llpcCompiler.cpp

```diff
@@ -1230,12 +1243,153 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+        if (shaderInfoEntry->pModuleData && isShaderStageInMask(stage, partStageMask)) {
+          // This is a shader to include in this part pipeline. Add the shader code and options to the hash.
+          PipelineDumper::updateHashForPipelineShaderInfo(stage, shaderInfoEntry, /*isCacheHash=*/true, &hasher,
+                                                          /*isRelocatableShader=*/false);
+        }
+      }
+    }
+    // Add applicable pipeline state to the hash. (This uses getShaderStageMask() to decide which parts of the
+    // state are applicable.)
+    context->getPipelineContext()->setPipelineState(/*pipeline=*/nullptr, /*hasher=*/&hasher, /*unlinked=*/false);
+
+    // If we are doing the pre-rasterization part pipeline, the linker's pipeline object contains the FS input mapping
+    // state. We need to include that state in the hash.
+    if (partPipelineStage == PartPipelineStagePreRasterization) {
+      assert(elfLinker);
+      StringRef fsInputMappings = elfLinker->getFsInputMappings();
+      hasher.Update(reinterpret_cast<const uint8_t *>(fsInputMappings.data()), fsInputMappings.size());
+    }
+
+    // Finalize the hash, and look it up in the cache.
+    MetroHash::Hash partPipelineHash = {};
+    hasher.Finalize(partPipelineHash.bytes);
+    CacheAccessor cacheAccessor(context, partPipelineHash, getInternalCaches());
+    if (cacheAccessor.isInCache()) {
+      LLPC_OUTS("Cache hit for stage " << getPartPipelineStageName(partPipelineStage) << ".\n");
+
+      // Mark the applicable entries in stageCacheAccesses.
+      for (ShaderStage shaderStage : maskToShaderStages(partStageMask)) {
+        stageCacheAccesses[shaderStage] =
+            cacheAccessor.hitInternalCache() ? CacheAccessInfo::InternalCacheHit : CacheAccessInfo::CacheHit;
+      }
+      // Get the ELF from the cache.
+      partPipelineElf = llvm::StringRef(static_cast<const char *>(cacheAccessor.getElfFromCache().pCode),
+                                        cacheAccessor.getElfFromCache().codeSize);
+    } else {
+      LLPC_OUTS("Cache miss for stage " << getPartPipelineStageName(partPipelineStage) << ".\n");
+
+      // Compile the part pipeline.
+      // Keep the PipelineShaderInfo structs for the shaders we are including.
+      llvm::SmallVector<const PipelineShaderInfo *, 4> partPipelineShaderInfo;
+      for (const PipelineShaderInfo *oneShaderInfo : shaderInfo) {
+        if (oneShaderInfo && isShaderStageInMask(oneShaderInfo->entryStage, partStageMask))
+          partPipelineShaderInfo.push_back(oneShaderInfo);
+      }
+      auto otherPartPipeline = (partPipelineStage == PartPipelineStagePreRasterization) ? &*elfLinkerPipeline : nullptr;
+      result = buildPipelineInternal(context, partPipelineShaderInfo, PipelineLink::PartPipeline, otherPartPipeline,
+                                     &partPipelineBuffers[partPipelineStage], stageCacheAccesses);
+      if (result != Result::Success)
+        break;
```

Comment:
nit: can we `return` here instead of `break`ing? If yes, we could get rid of the `result` variable.